### ISSUE TITLE
[Request] Trakt Progress page

### ISSUE THREAD

**Issue by hoshsadiq** [elgatito/plugin.video.elementum#46]
On Trakt you have a TV show progress section (https://trakt.tv/users/<user\>/progress/watched/activity) that tells you what the next episode of your unfinished shows is. Would be cool to have this. I think this has already been implemented in this PR: https://github.com/scakemyer/quasar/pull/40 but I imagine some changes will be required as I've seen some major refactoring in elementum.

**Comment by burekas7** [comment]
But look on my  comments here:
https://github.com/scakemyer/plugin.video.quasar/issues/580

The best implement of this category is in SALTS.

**Comment by hoshsadiq** [comment]
I saw that thread too. Regarding the difference you asked about on that thread. On the trakt website by the default shows by default shows all shows. Including the ones you have finished etc. It's possible to filter these out. I'm not sure how salts works, but in Exodus, shows that you were up to date on (i.e. next episode is in the future or there is no next episode), it would filter them out. Otherwise the show was shown, and if you opened it, it would list all next shows to watch. If after the next episode to watch, there were more episodes and they were not aired yet, they would be greyed out. I think that's more than enough usability.

**Comment by hoshsadiq** [comment]
Actually, I think you're mistaking this feature request with what was already implemented. The PR I linked to is what you were asking for.

**Comment by elgatito** [comment]
You are welcome to try 0.0.35 with Progress page.

**Comment by hoshsadiq** [comment]
When I go to TV Shows > Trakt, the following notification comes up:
```
<bold>Elementum</bold>
Not Found
```

[kodi.log](https://paste.ee/p/IKrpL)

**Comment by elgatito** [comment]
Oh.  You have cached list. Can delete cache.db or wait

**Comment by burekas7** [comment]
Thanks @elgatito !
When I asked for it in Quasar, he didn't implement it well.

But I have some improvement requests (Regarding the SALTS addon that also implemented it):
1. Colorize the air date and the aired/not aired episode title in different color for distinction will be nice also. For example:
* This is how it shown in Elementum:
![progresselementum](https://user-images.githubusercontent.com/13424257/35400021-a0151ffc-01ed-11e8-93cc-80fb5e4d0d43.png)

* And this is what I asked:
![progresssalts](https://user-images.githubusercontent.com/13424257/35400030-a64c8c2a-01ed-11e8-96b9-0672dfa577b5.png)

2. Add an options for sorting this progress list:
- By Date
- By Title (Name)
- Recently collected

3. Add an option for setting the date format as you like (For example: %d-%m-%y or %y-%m-%d)

Thanks.

**Comment by hoshsadiq** [comment]
Great! Works. Life will be much better with this. Thank you so much!

I think one other cool thing would be to separate episodes into a new page. I.e. when going into `Progress`, you see each individual show on your progress list, and what the next episode would be. Then when you click it, it shows a list of all the unwatched episodes. The video in "new episodes" is good enough IMO.

**Comment by hoshsadiq** [comment]
By the way, I'm happy to help with stuff, however, I need some guidance on setting up a dev environment as I mentioned in #48

**Comment by burekas7** [comment]
@hoshsadiq 
I think it's good as it now, all the 'next episodes' in the same page.
If you want to see all the next episodes separated by TV shows just go to your collection. 'Progress' should show only the next episode for each TV Show.

**Comment by elgatito** [comment]
> I think one other cool thing would be to separate episodes into a new page. I.e. when going into Progress, you see each individual show on your progress list, and what the next episode would be. Then when you click it, it shows a list of all the unwatched episodes. The video in "new episodes" is good enough IMO.

You can open "History" page, which shows the same order of the same shows, which you can enter and select the episode and so on.

**Comment by elgatito** [comment]
> By the way, I'm happy to help with stuff, however, I need some guidance on setting up a dev environment as I mentioned in #48

I remember. I will make some docs when I finish with basic stuff and it becomes more stable.

**Comment by elgatito** [comment]
> Colorize the air date and the aired/not aired episode title in different color for distinction will be nice also. 

I will not color anything. People have different skins and colors. Forcing colors is a bad idea. 
I don't like the idea of coloring anything in the lists at all.

> Add an options (In the Settings) for sorting this progress list

We get list of shows from Trakt sorted by "last watched". Custom sorting of items is done by Kodi, not the plugin.

> Why in Elementum it doesn't take the episode's name like it does in SALTS image? "Episode 1" instead of "Eden".

We get information about names from TMDB. You can edit information there, it's open for changes.

**Comment by burekas7** [comment]
* OK, ignore the colors thing (If you can add it as an option in Settings it could be nice).
* But, can you add at least the aired date in the text there with this date format of %d-%m-%y? (Like in the second example?)
* Regarding the sorting, Kodi has only two option 'Date' and 'Genere' and this is no what I meant.
And by the way the 'Date' sorting in kodi doesn't work here well.
In SALTS it is done by the addon and not by kodi.
* And regarding the episode name, I think SALTS take the information from Trakt, so this maybe explain the difference.
- Can you add an option of where to take the info from, TMDB or Trakt?
- What is the purpose actually of this setting: "Use Trakt information if available"?

**Comment by elgatito** [comment]
> What is the purpose actually of this setting: "Use Trakt information if available"?

Use original Trakt movies/episodes when browsing Trakt lists, by default we always use TMDB for information.

**Comment by burekas7** [comment]
> Use original Trakt movies/episodes when browsing Trakt lists, by default we always use TMDB for information.

I have enabled it, but the info is still taken from TMDB instead of Trakt.
The episode name in 'Progress' (Trakt) is still "Episode 1" instead of "Eden", also the thumbnails, fanart and dates are not from Trakt.

**Comment by hoshsadiq** [comment]
Both history and collections is not what I'm looking for. Progress is meant to hide everything that you've already fully watched. I've realised today that this is not the case. It shows future episodes.

**Comment by burekas7** [comment]
I think instead of "Progress", it should be called "My next episodes".

@hoshsadiq 
He implemented exactly what you said on your first comment, and it also like what I wanted.
I also don't understand what do you mean.
Why do you need the lone way of TV show and then the episode?
Why do not see all the next episodes in one place (page) like here? which is better.

**Comment by hoshsadiq** [comment]
Because you're not necessarily up to date with the shows. If you have a show that is on season 2 episode 5 for example, and you've only seen season 1, you've got 5 episodes left. The way it's implemented, you watch an episode, then you have to wait for it to refresh (which actually, oddly takes longer than all other menus) then you can watch the next episode.

2 more things
IMO the list should be ordered by airdate of the next episode you're meant to watch. 
Finally, I'm confused why future episodes should be included in the list. You can't do anything with them since they won't be out yet, so why have it in the list?

**Comment by elgatito** [comment]
> Progress is meant to hide everything that you've already fully watched.

You don't see a show when there is no unwatched episode in the queue.

> IMO the list should be ordered by airdate of the next episode you're meant to watch.

No. If you want to start watching old show - you won't see it in the list because air date was too long ago. Other side you want to watch new show - you will see only old shows with earlier air dates.
I like the way it's on Trakt, and did the same way in Elementum. If you want some show - you see the next episode there in the first place, then goes the second-watched show and so on. I don't see any reason why I want to see list of watched shows but not wanting the next episode of that show.

If you want to see list of episodes on the show - it's what Kodi does for you in the library.

**Comment by burekas7** [comment]
I'm also agree with @elgatito , the content of this list is correct and this is how it should be.
It's ok to see unaired episodes also in this list, this is one of the advantages here, on this way you also know when your tv show comes from a break or renew.

One thing I agree is about the sorting issue as I mentioned before,
It should be sorted by air date by default.
It can has other option of sorting like this:
- Air date
- Recently watched
- Name

* And it would be nice to mention the air date in the start of the title, because all the point in this list is to see the air dates.

Thanks.

**Comment by hoshsadiq** [comment]
> You don't see a show when there is no unwatched episode in the queue.

Sorry, I think you misunderstood me, the next sentence was part of it too. It seems that episodes that are not aired yet (but are listed on trakt as a future episode), are also listed in progress. For example, I have The Flash in my watch list, which currently I'm up to date on. Last episode aired was 4x11, which aired on Jan 24. When I go to progress, it shows The Flash with next episode being 4x11. Now I watched that episode yesterday, and now it shows 4x12, which hasn't been aired yet (airs tomorrow), and is therefore not possible to watch. These should be filtered out from the list.

**Comment by hoshsadiq** [comment]
I just looked at the docs for that endpoint. The documentation says:

> **Note**: Only aired episodes are used to calculate progress. Episodes in the future or without an air date are ignored.

However, when doing the request I get the following:
```
$ curl --header "Content-Type: application/json" --header "Authorization: Bearer 9c44266491ebfece7b061cee0557c32da9728213b71af9329d94551322474343" --header "trakt-api-version: 2" --header "trakt-api-key: 2f911cee953f0af7833191d2b929e9a842bf8752e6b1afb458c8ff9ffc1d2c85" 'https://api.trakt.tv/shows/the-flash-2014/progress/watched?extended=full' | jq

{
  "aired": 80,
  "completed": 80,
  "last_watched_at": "2018-01-25T23:57:26.000Z",
  "seasons": [
    ...
    {
      "number": 4,
      "aired": 11,
      "completed": 11,
      "episodes": [
        ...
        {
          "number": 10,
          "completed": true,
          "last_watched_at": "2018-01-17T20:57:15.000Z"
        },
        {
          "number": 11,
          "completed": true,
          "last_watched_at": "2018-01-25T23:57:26.000Z"
        }
      ]
    }
  ],
  "hidden_seasons": [],
  "next_episode": {
    "season": 4,
    "number": 12,
    "title": "Honey, I Shrunk Team Flash",
    "ids": {...},
    "number_abs": 81,
    "overview": "...",
    "rating": 7.53846,
    "votes": 13,
    "comment_count": 1,
    "first_aired": "2018-01-31T01:00:00.000Z",
    "updated_at": "2018-01-30T18:29:38.000Z",
    "available_translations": [...],
    "runtime": 44
  },
  "last_episode": {
    "season": 4,
    "number": 11,
    "title": "The Elongated Knight Rises",
    "ids": {...},
    "number_abs": 80,
    "overview": "...",
    "rating": 7.68019,
    "votes": 1676,
    "comment_count": 8,
    "first_aired": "2018-01-24T01:00:00.000Z",
    "updated_at": "2018-01-30T20:15:16.000Z",
    "available_translations": [...],
    "runtime": 42
  }
}
```

with the following headers
```
HTTP/2 200
date: Tue, 30 Jan 2018 20:50:20 GMT
content-type: application/json; charset=utf-8
last-modified: Thu, 25 Jan 2018 23:57:26 GMT
```

Based on the docs, `next_episode` should be null, as `next_episode.first_aired` is in the future (based on `date` in the headers.

**Comment by burekas7** [comment]
> which hasn't been aired yet (airs tomorrow), and is therefore not possible to watch. These should be filtered out from the list.

I don't think so. It's good to know in this list also if there are future episodes and their air dates.
There is a date info attached to each item so you should know if you can play it or not.
If it so annoying you, he can add a new setting in the settings to filter out the unaired episodes. But I think it's necessary that they are also there. 

This is an a example of the settings in SALTS addon regarding the My Next Episode feature ('Progress' here)
![salts](https://user-images.githubusercontent.com/13424257/35592239-46ffcf96-0604-11e8-9ee2-92d3e7fa2b39.png)

**Comment by elgatito** [comment]
> Based on the docs, next_episode should be null, as next_episode.first_aired is in the future (based on date in the headers).

Yes, it should have been set to null. I will add a check there to skip those items.

> This explains my previous comment. I'm trying to find a place where I can report this as a bug to see if their devs can look into it.

Not sure where to report. There is a Google+ community for Trakt API, but not sure if that is the right place.

Still it's so much slow, you need to get list of shows, then request each show and only after that you know how much of them have available episodes. One more thing is that Trakt can have empty TMDB link or link to wrong TMDB item, so we fetch the show and it does not exist and we skip it's output, so the list is again broken (not filling right number of items).

**Comment by burekas7** [comment]
For me it's not a bug so I hope they won't fix that :)
There are some people like me that wants to know also about future episodes and their airdates, instead of checking all the time all kind of sites with TV Shows renew/cancel lists.

**Comment by hoshsadiq** [comment]
@elgatito 
So I got a reply:
> Next episode will be populated, even if it airs in the future. The note in the docs is saying that only aired episodes are used in the progress count and calculations. `next_episode` is just for informational use in your app and isn't related to the calculations. In this case, you've watched 100%, so `next_episode` would show the next episode in the future (as long as we have that info) so you know there will be something to watch.

> I will add a check there to skip those items.
Thanks. This is very helpful.

> Still it's so much slow, you need to get list of shows, then request each show and only after that you know how much of them have available episodes.

Agreed. I am happy to contact the devs and request this endpoint if you'd like. It would be nice to improve the speed of this page.

Lastly, apparently the G+ community you mentioned before is the right place for such questions. I've asked if it makes more sense to have a Github community.

**Comment by burekas7** [comment]
@elgatito 
> Yes, it should have been set to null. I will add a check there to skip those items.
Can you add it as an option in the Settings? Because there are some other that want it as it's working now (Without skip those items)

**Comment by hoshsadiq** [comment]
@burekas7 maybe I'm misunderstanding your use case, but going to `TV Shows > Calendars > My Shows`, shows you all episodes that are upcoming in much the same way as the progress page. Is this perhaps what you're looking for?

**Comment by burekas7** [comment]
@hoshsadiq 
No.
This path shows me only one episode there, this is what was done wrong in Quasar. He didn't implemented what I meant.
Here in 'Progress' is exactly what I meant to (Like it is done in SALTS addon and calls there as 'My Next Episode'), when it shows **all** the next/future episodes (Those that already aired and those that still not aired yet but has future air date).
I don't know why it called here 'Progress', it should be called 'My next episodes'.

**Comment by elgatito** [comment]
>> I will add a check there to skip those items.
> Thanks. This is very helpful.

I have used the existing "show unaired episodes" options to influence listing episodes in Progress with "Aired" being after now. 

I would be fast if we just use paging and split to usual number per page. But. We are checking each trakt show with TMDB database and some items should be dropped off the list, then we check episode, which also can be not in the TMDB database, and then it can be in the future, so also drop off, so you see you are getting less then required items per page. If it's the first page - it's fine, if not first - it's too ugly.  
Bad we can't get multiple episodes/shows/movies by one request :)

**Comment by hoshsadiq** [comment]
@elgatito I made a feature request on their G+ page: https://plus.google.com/104974452270599315078/posts/QpHJJ7y7n3t

**Comment by elgatito** [comment]
@hoshsadiq I have totally loose the point of this issue.

Can you describe how you see the *desired* progress page and what should be there? 
Now it's the same as Trakt's Progress page where you see Next episodes for each show that you watch, sorted by last watched time in descending order. ( https://trakt.tv/users/username/progress )

**Comment by hoshsadiq** [comment]
Yes, sorry, so I think I neglected to mention that I always have the below filter on. I think adding the additional airdate as @burekas7 mentioned, and hiding the items that have future airdates is what my I had in mind. Having those two imo would be enough for this issue to be closed

![image](https://user-images.githubusercontent.com/600045/35866508-4732eb42-0b4f-11e8-9ddd-cf1aa8828d82.png)

**Comment by elgatito** [comment]
Changes in v0.0.54:

- Added settings for colors, date format, sorting of items on the progress page.
- Refactored to use last activities, so now it's not too fast for the first time, then it should be quite fast.

Closing this one. Let it be a separate issue if there is something else to do with the progress page.
